### ISSUE TITLE
Extend schedule if not reached target SOC at end of schedule

### ISSUE THREAD

**Issue by AnderssonPer** [jonasbkarlsson/ev_smart_charging#346]
**Background**
My grid provider has recently introduced a fee based on peak power consumption. Charging at full power during one hour is way more expensive than charging at lower speed for several hours. The fee is also significant higher during daytime (06-23) than during night (23-06). So to get the lowest possible fees you need to charge slowly during nights (while still trying to charge during the hours where the electricity is cheapest). This is handled nicely by the "Charge start/completion time" settings.

To further complicate things I have a SPA, which isn't easily controllable. It's set to heat the water during the night hours, but depending on the outside temperature etc it can take anything from 45 minutes to several hours. It can start any time during the night. To lower my energy peaks I have an automation that stops the EV charging when the SPA starts it's heating, and then reactivate it when the SPA is done (and if sensor.ev_smart_charging_charging is "on").

My problem is that if the EV is scheduled to charge between eg 01-04, and the SPA decides to heat it's water for 2 hours during that time slot, the EV charging isn't finished at 4 when the schedule ends. Since I have "switch.ev_smart_charging_keep_charger_on" set to off (to avoid expensive charging during day time), this means that the EV charging is suspended even though it has not reached the target SOC and there is more "night time" available before the day start.

**Proposal**
Either add a "switch.ev_smart_charging_keep_charger_on_but_only_until_charging_completion_time" or make the integration recalculate the schedule when it reaches it's end time and the target SOC hasn't been reached and there is time to extend the schedule without exceeding the "Charging completion time".

**Comment by AnderssonPer** [comment]
Hm, I thought that I tested it and found out that it didn't work. I will try it again during the weekend and get back with the result.

**Comment by AnderssonPer** [comment]
You are right, it actually extends the schedule. This night it was scheduled to charge between 23 and 02, but as seen in the image it was extended. Sorry for false alarm. 🙂

But one detail in the image makes me a bit confused (and I think this might be the reason why I thought that we schedule wasn't extended). 
The vehicle seems to stop updating the SOC level between 03 and 03:58, and then sets it to 79%. The target SOC is 80%, but the Charging is aborted at 04 anyways. Is this expected?

It also looks like the vehicle suspends the charging at 03:58 by it self, but since it still sends SOC 79% I guess that we should still try to have the charging active?

![Screenshot_20241215-081800](https://github.com/user-attachments/assets/3b24d343-45c8-46c3-9e89-ce02b68e1fa4)

**Comment by ronlimon** [comment]
Strong +1 — a concrete use-case from our setup: our charger is 3-phase but occasionally falls back to 1-phase, and a cold battery pack can throttle acceptance, so a session sometimes under-delivers and the car ends the schedule below target with no recovery until the next plan. "Extend the schedule if target SOC isn't reached by the end" would close that gap.

This is planner-core, so the design is really your call, @jonasbkarlsson — but we're happy to provide input, edge-cases, or testing whenever it's useful.

**Comment by jonasbkarlsson** [comment]
I think there are two error cases.
1) The integration does not recalculate, even if there is more time until the completion time. 
2) The charging is slower than expected, so at the completion time, charging is not completed.

Which ones of these are you observing?

Apparently, I have seen 1) myself, according to my comment on Feb 22, 2025, if not charging was done at all. The original issue is also 1), I think, but if some charging was done. And for that sub-case, I think the integration has a function to recalculation, but I need to check more in detail.

@ronlimon, your case is 2). One possible solution to this is to turn on `switch.ev_smart_charging_keep_charger_on`. Would that be ok for you?
